modules/gapi/src/streaming/gstreamer/gstreamersource.cpp

```diff
@@ -0,0 +1,374 @@
+            result = retrieveFrame(mat);
+            if (result) {
+                data = mat;
+            }
+            break;
+        }
+        default: {
+            GAPI_Assert(false && "Unsupported output type for GStreamerSource!");
+        }
+    }
+
+    if (result) {
+        data.meta[cv::gapi::streaming::meta_tag::timestamp] = computeTimestamp();
+        data.meta[cv::gapi::streaming::meta_tag::seq_id]    = m_frameId++;
+    }
+
+    return result;
+}
+
+GMetaArg GStreamerSource::Priv::descr_of() noexcept
+{
+    // Prepare frame metadata if it wasn't prepared yet.
+    prepareVideoMeta();
+
+    switch(m_outputType) {
+        case GStreamerSource::OutputType::FRAME: {
+            return GMetaArg { m_mediaFrameMeta };
+        }
+        case GStreamerSource::OutputType::MAT: {
+            return GMetaArg { m_matMeta };
+        }
+        default: {
+            GAPI_Assert(false && "Unsupported output type for GStreamerSource!");
+        }
+    }
+}
+
+void GStreamerSource::Priv::configureAppsink() {
+    // NOTE: appsink element should be configured before pipeline launch.
+    GAPI_Assert(!m_pipeline->isPlaying());
+
+    // TODO: is 1 single buffer really high enough?
+    gst_app_sink_set_max_buffers(GST_APP_SINK(m_appsink.get()), 1);
+
+    // Do not emit signals: all calls will be synchronous and blocking.
+    gst_app_sink_set_emit_signals(GST_APP_SINK(m_appsink.get()), FALSE);
+
+    GStreamerPtr<GstCaps> nv12Caps(gst_caps_from_string(
+        "video/x-raw,format=NV12;video/x-raw(memory:DMABuf),format=NV12")); /* transfer full */
+
+    GStreamerPtr<GstPad> appsinkPad(gst_element_get_static_pad(m_appsink, "sink")); /* transfer full */
+    GStreamerPtr<GstCaps> peerCaps(gst_pad_peer_query_caps(appsinkPad, NULL)); /* transfer full */
+    if (!gst_caps_can_intersect(peerCaps, nv12Caps)) {
+        throw std::logic_error("appsink element can only consume video-frame in NV12 "
+                               "format in GStreamerSource");
+    }
+
+    gst_app_sink_set_caps(GST_APP_SINK(m_appsink.get()), nv12Caps);
+
+    gst_pad_add_probe(appsinkPad, GST_PAD_PROBE_TYPE_QUERY_DOWNSTREAM, appsinkQueryCallback,
+                      NULL, NULL);
+}
+
+void GStreamerSource::Priv::prepareVideoMeta()
+{
+    if (!m_isMetaPrepared) {
+        m_pipeline->completePreroll();
+
+        GStreamerPtr<GstSample> prerollSample(
+#if GST_VERSION_MINOR >= 10
+            gst_app_sink_try_pull_preroll(GST_APP_SINK(m_appsink.get()), 5 * GST_SECOND)); /* transfer full */
+#else // GST_VERSION_MINOR < 10
+            // TODO: This function may cause hang with some pipelines, need to check whether these
+            // pipelines are really wrong or not?
+            gst_app_sink_pull_preroll(GST_APP_SINK(m_appsink.get()))); /* transfer full */
+#endif // GST_VERSION_MINOR >= 10
+        GAPI_Assert(prerollSample != nullptr);
+
+        GstCaps* prerollCaps(gst_sample_get_caps(prerollSample)); /* transfer none */
```

Comment:
Do we need to wrap it in GStreamerPtr?

Reply:
```transfer none``` in comment means that accepting code of ```gst_sample_get_caps``` function return value will not own memory for this value and shouldn't release it